doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,416 @@
+Materialize) if data fails to decode, as the upstream systems enforce schemas.
+But Kafka does not guarantee schema enforcement, so it is common for bad data to
+slip into a Kafka topic.
+
+Because Kafka sources are the most prone to bad data, we propose to pursue a
+Kafka source-specific solution for handling source decoding errors.
+
+### User experience
+
+#### Dead letter queue
+
+We propose to introduce a [dead letter queue][dlq] (DLQ) for Kafka sources.
+
+Intuitively, the DLQ provides a running log of all decoding errors encountered
+by the source. Materialize users can monitor the count of errors in the DLQ and
+set up alerts whenever the count increases. Users can filter on `mz_timestamp`
+and/or `timestamp` to eliminate old errors that are no longer relevant. (Using
+the former column filters by ingestion time, while filtering on the latter
+column filters by the time the message was produced to Kafka.)
+
+Concretely, we propose to add a `REDIRECT ERRORS` option to `CREATE SOURCE` that
+specifies the name of a DLQ table in which to emit information about undecodable
+messages:
+
+```sql
+CREATE SOURCE src TO KAFKA CONNECTION kconn ... WITH (
+    REDIRECT ERRORS = dlq
+)
+```
+
+The specified DLQ table (`dlq` in the example above) must not exist before the
+`CREATE SOURCE` command is executed. Materialize will automatically create the
+DLQ table with the following Kafka-specific structure.
+
+ Name           | Type           | Nullable | Description
+----------------|----------------|----------|------------
+ `mz_timestamp` | `mz_timestamp` | No       | The logical timestamp that the message was reclocked to.
+ `error`        | `text`         | No       | The text of the decoding error.
+ `error_code`   | `text`         | No       | The code of the decoding error.
+ `partition`    | `integer`      | No       | The partition of the Kafka message that failed to decode.
+ `offset`       | `bigint`       | No       | The offset of the Kafka message that failed to decode.
+ `timestamp`    | `timestamp`    | No       | The timestamp of the Kafka message that failed to decode.
+ `key`          | `bytea`        | Yes      | The key bytes of the Kafka message that failed to decode.
+ `value`        | `bytea`        | Yes      | The value bytes of the Kafka message that failed to decode.
+
+The DLQ table works like a subsource. It can be queried via `SELECT`, subscribed
+to via `SUBSCRIBE`, and used as the upstream relation for a Kafka sink.
+
+The DLQ table is append only. Allowing users to specify a retention policy on
+the DLQ table is left to future work. Users cannot issue `INSERT`, `UPDATE`, or
+`DELETE` commands against the DLQ[^2].
+
+When a DLQ table is attached to a source, no errors are emitted to the source's
+*errs stream* (see the [error handling section][error-handling-docs] of the
+`compute::render` module).
+
+**Open questions:**
+
+  * Is there a better name for the option than `REDIRECT ERRORS`?
```

Comment:
Hm, is the equivalent Snowpipe option not `ERROR_INTEGRATION`? `ON_ERROR` seems to be about whether to `CONTINUE` or to `SKIP_FILE`.

Reply:
Aren't they related? At least my understanding is that `ERROR_INTEGRATION` relies on setting up a notification integration, which is only triggered if `ON_ERROR` is set to `SKIP_FILE`. To query the errors, you then have to use the `NOTIFICATION_HISTORY` table function. But this is for Snowpipe (batch).

For the (Snowpipe Streaming-based) Kafka connector, they're also following a DLQ approach ([docs](https://docs.snowflake.com/user-guide/data-load-snowpipe-streaming-kafka#error-handling-and-dlq-properties)).

Reply:
Yes, they're related in Snowflake! What I'm getting at is that I don't think `REDIRECT ERRORS` option proposed here maps directly to the `ON_ERROR` option. `REDIRECT ERRORS` seems closer to the `ERROR_INTEGRATION` option than the `ON_ERROR` option.

If we wanted the to to be more `ON_ERROR`-like, we could do `ON ERROR = {RECORD | SEND TO <name>}`, where `RECORD` means "record it in this source" like we do presently, and `SEND TO <name>` means "do the DLQ thing."

Reply:
What I meant is that...it doesn't map to either because error logging in Snowpipe uses a different logic, whereas in Snowpipe Streaming `REDIRECT ERRORS` is directly equivalent to [`errors.deadletterqueue.topic.name`](https://docs.snowflake.com/user-guide/data-load-snowpipe-streaming-kafka#error-handling-and-dlq-properties)!